nifi-nar-bundles/nifi-standard-services/nifi-distributed-cache-services-bundle/nifi-distributed-cache-server/src/test/java/org/apache/nifi/distributed/cache/server/map/TestDistributedMapServerAndClient.java

```diff
@@ -298,6 +301,45 @@ public void testLimitServiceReadSize() throws InitializationException, IOExcepti
         }
     }
 
+    @Test
+    public void testIncompleteHandshakeScenario() throws InitializationException, IOException {
+        // Default port used by Distributed Server and Client
+        final int port = 4557;
+
+        // This is used to simulate a DistributedCacheServer that does not complete the handshake response
+        final ServerSocketChannel serverSocketChannel = ServerSocketChannel.open();
+        serverSocketChannel.configureBlocking(true);
+        serverSocketChannel.bind(new InetSocketAddress(port));
+        final Thread thread = startServerSocket(serverSocketChannel);
+
+        DistributedMapCacheClientService client = new DistributedMapCacheClientService();
+        MockControllerServiceInitializationContext clientInitContext = new MockControllerServiceInitializationContext(client, "client");
+        client.initialize(clientInitContext);
+
+        final Map<PropertyDescriptor, String> clientProperties = new HashMap<>();
+        clientProperties.put(DistributedMapCacheClientService.HOSTNAME, "localhost");
+        clientProperties.put(DistributedMapCacheClientService.PORT, String.valueOf(port));
+        clientProperties.put(DistributedMapCacheClientService.COMMUNICATIONS_TIMEOUT, "1 sec");
+
+        MockConfigurationContext clientContext = new MockConfigurationContext(clientProperties, clientInitContext.getControllerServiceLookup());
+        client.onEnabled(clientContext);
+
+        final Serializer<String> valueSerializer = new StringSerializer();
+        final Serializer<String> keySerializer = new StringSerializer();
+        final Deserializer<String> deserializer = new StringDeserializer();
+
+        try {
+            client.getAndPutIfAbsent("testKey", "test", keySerializer, valueSerializer, deserializer);
+            fail("Client operation should have failed due to it not being connected to a DistributedCacheServer");
+        } catch (IOException e) {
+            // Verify cause of exception was handshake completion timeout
+            assertEquals("Handshake timed out before completion.", e.getCause().getMessage());
+        } finally {
+            thread.interrupt();
+            serverSocketChannel.close();
+        }
```

Comment:
Instead of try-catch with `fail()`, it would be better to use `assertThrows()`. It is also best to avoid check for the exact error message.